File: api/src/main/java/com/sahilkumar/api/billing/Bill.java
package com.sahilkumar.api.billing;

import com.sahilkumar.api.common.BaseEntity;
import com.sahilkumar.api.customer.Customer;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.math.BigDecimal;
import java.time.Instant;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "bills", uniqueConstraints = @UniqueConstraint(
        name = "uq_bill_per_customer_period",
        columnNames = {"customer_id", "period_year", "period_month"}))
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Bill extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @Column(name = "period_month", nullable = false)
    private int periodMonth;

    @Column(name = "period_year", nullable = false)
    private int periodYear;

    @Column(name = "total_liters", nullable = false, precision = 12, scale = 3)
    private BigDecimal totalLiters;

    @Column(name = "total_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal totalAmount;

    @Column(name = "paid_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal paidAmount;

    @Column(name = "remaining_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal remainingAmount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private BillStatus status;

    @Column(name = "generated_at", nullable = false)
    private Instant generatedAt;
}

File: android/app/src/main/java/com/sahilkumar/dfms/model/Bill.kt
package com.sahilkumar.dfms.model

import java.math.BigDecimal

data class BillResponse(
    val id: String,
    val customerId: String,
    val customerName: String,
    val periodMonth: Int,
    val periodYear: Int,
    val totalLiters: BigDecimal,
    val totalAmount: BigDecimal,
    val paidAmount: BigDecimal,
    val remainingAmount: BigDecimal,
    val status: BillStatus,
    val generatedAt: String?,
)

data class GenerateBillsRequest(
    val month: Int,
    val year: Int,
)

data class PaymentResponse(
    val id: String,
    val billId: String,
    val amount: BigDecimal,
    val paymentMethod: PaymentMethod,
    val paymentDate: String,
    val reference: String?,
    val notes: String?,
)

data class PaymentRequest(
    val amount: BigDecimal,
    val paymentMethod: PaymentMethod,
    val paymentDate: String,
    val reference: String?,
    val notes: String?,
)
